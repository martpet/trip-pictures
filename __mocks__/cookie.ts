export default {
  serialize: vi.fn().mockName('serialize').mockReturnValue('dummyCookie'),
  parse: vi.fn().mockName('parse'),
};
