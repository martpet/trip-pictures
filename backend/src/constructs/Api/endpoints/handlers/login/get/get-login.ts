import {
  APIGatewayProxyHandlerV2,
  apiPaths,
  ApiQueryStrings,
  authPaths,
  cookie,
  cookieName,
  EnvVars,
  errorResponse,
  OauthCookieProps,
  StatusCodes,
} from 'lambda-layer';

import { generateOauthRandoms } from './generateOauthRandoms';

export const handler: APIGatewayProxyHandlerV2 = async ({ queryStringParameters }) => {
  const { clientId, authDomain, loginCallbackUrl } = process.env as EnvVars<
    '/login',
    'GET'
  >;

  const { provider } = Object(queryStringParameters) as ApiQueryStrings<'/login'>;

  if (!clientId || !authDomain || !loginCallbackUrl) {
    return errorResponse('1rQkj3kpp4');
  }

  if (!provider) {
    return errorResponse('6gUyCQQT7Z', { statusCode: StatusCodes.BAD_REQUEST });
  }

  const { idTokenNonce, stateNonce, codeChallenge, codeVerifier } =
    await generateOauthRandoms();

  const cognitoAuthUrl = new URL(`https://${authDomain}${authPaths.authorize}`);

  const cognitoAuthUrlParams = new URLSearchParams({
    identity_provider: provider,
    client_id: clientId,
    redirect_uri: loginCallbackUrl,
    response_type: 'code',
    state: stateNonce,
    nonce: idTokenNonce,
    code_challenge_method: 'S256',
    code_challenge: codeChallenge,
  });

  cognitoAuthUrl.search = cognitoAuthUrlParams.toString();

  const oauthCookieProps: OauthCookieProps = {
    stateNonce,
    idTokenNonce,
    codeVerifier,
  };

  const oauthCookie = cookie.serialize(
    cookieName('oauth'),
    JSON.stringify(oauthCookieProps),
    {
      httpOnly: true,
      secure: true,
      sameSite: 'lax',
      path: apiPaths['login-callback'],
    }
  );

  return {
    statusCode: StatusCodes.MOVED_TEMPORARILY,
    headers: { location: cognitoAuthUrl.href },
    cookies: [oauthCookie],
  };
};
