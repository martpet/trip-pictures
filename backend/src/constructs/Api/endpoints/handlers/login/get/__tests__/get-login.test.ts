import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import cookie from 'cookie';

import {
  itCalls,
  itHasEnvVars,
  itHasQueryStrings,
  itResolves,
} from '~/constructs/Api/utils';

import { handler } from '../get-login';

vi.mock('cookie');
vi.mock('~/constructs/Api/utils/errorResponse');
vi.mock('../generateOauthRandoms');

process.env.clientId = 'dummyClientId';
process.env.authDomain = 'dummyAuthDomain';
process.env.loginCallbackUrl = 'dummyLoginCallbackUrl';

const args = [
  {
    queryStringParameters: {
      provider: 'dummyProvider',
    },
  },
] as unknown as Parameters<APIGatewayProxyHandlerV2>;

describe('"get-login" handler', () => {
  itHasQueryStrings(['provider'], handler, args);
  itHasEnvVars(['authDomain', 'loginCallbackUrl', 'clientId'], handler, args);
  itCalls(cookie.serialize, handler, args);
  itResolves(handler, args);
});
