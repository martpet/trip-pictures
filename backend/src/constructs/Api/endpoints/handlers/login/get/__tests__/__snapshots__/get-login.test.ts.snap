// Vitest Snapshot v1

exports[`"get-login" handler > calls "serialize" with correct args 1`] = `
[
  [
    "oauth",
    "{\\"stateNonce\\":\\"dummyStateNonce\\",\\"codeVerifier\\":\\"dummyCodeVerifier\\"}",
    {
      "httpOnly": true,
      "path": "/login-callback",
      "sameSite": "lax",
      "secure": true,
    },
  ],
]
`;

exports[`"get-login" handler > resolves with a correct value 1`] = `
{
  "cookies": [
    "dummyCookie",
  ],
  "headers": {
    "location": "https://dummyauthdomain/oauth2/authorize?identity_provider=dummyProvider&client_id=dummyClientId&redirect_uri=dummyLoginCallbackUrl&response_type=code&state=dummyStateNonce&nonce=undefined&code_challenge_method=S256&code_challenge=dummyCodeChallenge",
  },
  "statusCode": 302,
}
`;

exports[`"get-login" handler > when "authDomain" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "1rQkj3kpp4",
  ],
]
`;

exports[`"get-login" handler > when "authDomain" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-login" handler > when "clientId" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "1rQkj3kpp4",
  ],
]
`;

exports[`"get-login" handler > when "clientId" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-login" handler > when "loginCallbackUrl" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "1rQkj3kpp4",
  ],
]
`;

exports[`"get-login" handler > when "loginCallbackUrl" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`"get-login" handler > when "provider" query string parameter is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "6gUyCQQT7Z",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`"get-login" handler > when "provider" query string parameter is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;
